test/src/unit-cppapi-bitsort-filter.cc

```diff
@@ -0,0 +1,575 @@
+/**
+ * @file   unit-cppapi-bitsort-filter.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the C++ API for bitsort filter related functions.
+ */
+
+#include <random>
+#include <vector>
+
+#include "catch.hpp"
+#include "tiledb/common/common.h"
+#include "tiledb/sm/cpp_api/tiledb"
+
+using namespace tiledb;
+
+std::string bitsort_array_name = "cpp_unit_array";
+int bitsort_dim_hi = 10;
+int tile_extent = 4;
+
+// Defining distribution parameters.
+typedef typename std::uniform_int_distribution<uint64_t>
+    UnsignedIntDistribution;
+typedef typename std::uniform_int_distribution<int64_t> IntDistribution;
+typedef typename std::uniform_real_distribution<double> FloatDistribution;
+
+/**
+ * @brief Set the buffer with the appropriate dimensions for a 1D (10x) array.
+ *
+ * @tparam T The type of the dimension.
+ * @param x_dims The buffer that the function sets.
+ */
+template <typename T>
+void set_1d_dim_buffers(std::vector<T>& x_dims) {
+  for (int x = 0; x < bitsort_dim_hi; ++x) {
+    x_dims.push_back(static_cast<T>(x + 1));
+  }
+}
+
+/**
+ * @brief  Set the buffer with the appropriate dimensions for a 2D (10x10)
+ * array.
+ *
+ * @tparam T The type of the dimension.
+ * @param x_dims The first buffer that the function sets.
+ * @param y_dims The second buffer that the function sets.
+ */
+template <typename T>
```

Comment:
I changed it for when I thought it was needed. I thought it was needed for tiles where the type indicated info but it was not specified. For functions where the variable name gave the use of the template, I left it alone (e.g.)

```
  /**
   * Runs the input tiles for the input attribute through the filter pipeline.
   * The tile buffers are modified to contain the output of the pipeline.
   *
   * @param name The attribute/dimension the tiles belong to.
   * @param tile The tiles to be filtered.
   * @param dim_tiles The dimension tile array (used for BitSortFilter).
   * @return Status
   */

  template <typename SupportTileType>
  Status filter_tiles(
      const std::string& name,
      WriterTileVector* tiles,
      const std::vector<WriterTileVector*>& dim_tiles = {});

  /**
   * Runs the input tile for the input attribute/dimension through the filter
   * pipeline. The tile buffer is modified to contain the output of the
   * pipeline.
   *
   * @tparam T Type of the support argument passed to the filter pipeline.
   * @param name The attribute/dimension the tile belong to.
   * @param tile The tile to be filtered.
   * @param offsets_tile The offsets tile in case of a var tile, or null.
   * @param offsets True if the tile to be filtered contains offsets for a
   *    var-sized attribute/dimension.
   * @param offsets True if the tile to be filtered contains validity values.
   * @return Status
   */

  template <
      typename T,
      typename std::enable_if<!std::is_same<T, std::nullptr_t>::value>::type* =
          nullptr>
  Status filter_tile(
      const std::string& name,
      Tile* tile,
      T support_tile,
      bool offsets,
      bool nullable);
```

I changed the former one since there isn't an argument that tells us what T is for, but did not change the latter because it seemed redundant.

Reply:
@abi I think it will clarify the code in the .cc file and prevent readers from scrolling or going to the header when trying to understand the code.